pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/MmapMemory.java

```diff
@@ -0,0 +1,377 @@
+    }
+  }
+
+  @Override
+  protected void finalize()
+      throws Throwable {
+    if (!_closed) {
+      LOGGER.warn("Mmap section of " + _size + " wasn't explicitly closed");
+      close();
+    }
+    super.finalize();
+  }
+
+  private static class MapSection {
+    public static final MapSection EMPTY = new MapSection(0, () -> {
+    });
+    private final long _address;
+    private final UnmapFun _unmapFun;
+
+    public MapSection(long address, UnmapFun unmapFun) {
+      _address = address;
+      _unmapFun = unmapFun;
+    }
+
+    public long getAddress() {
+      return _address;
+    }
+
+    public UnmapFun getUnmapFun() {
+      return _unmapFun;
+    }
+  }
+
+  /**
+   * This is a factory method that can be used to create {@link MapSection}s.
+   *
+   * Each JVM may provide different method to map files in memory.
+   */
+  interface MapFun {
+
+    /**
+     * @param file The file to be mapped. If its length is lower than offset + size and the mode is not read only,
+     *            the file will be resized to that size.
+     * @param offset The offset in the file. Any positive value is valid, even if it is larger than the file size.
+     * @param size How many bytes to map.
+     * @throws IOException in several situations. For example, if the offset + size is larger than file length and the
+     * mode is read only or if the process doesn't have permission to read or modify the file.
+     */
+    MapSection map(File file, boolean readOnly, long offset, long size) throws IOException;
+
+    static MapFun find()
+        throws ClassNotFoundException, NoSuchMethodException {
+      List<Finder<? extends MapFun>> candidates = Lists.newArrayList(
+          new Map0Fun.ChronicleCore(),
+          new Map0Fun.Java11(),
+          new Map0Fun.Java17(),
+          new Map0Fun.Java20()
+      );
+
+      for (Finder<? extends MapFun> candidate : candidates) {
+        try {
+          return candidate.tryFind();
+        } catch (NoSuchMethodException | ClassNotFoundException | AssertionError e) {
+          // IGNORE
+        }
+      }
+      throw new NoSuchMethodException("Cannot find how to create memory map files in Java " + JavaVersion.VERSION);
+    }
+  }
+
+  /**
+   * As defined by POSIX, the map0 method requires that the offset is page aligned. Failing to do that may produce
+   * segfault errors. This interface is a {@link MapFun} that does some sanitation before calling the map method.
+   * They include:
+   * <ul>
+   *   <li>Grow the file if the last mapped byte is larger than the file length.</li>
+   *   <li>Align the offset with the previous page. This means that we need to correct the actual mapped address.</li>
+   * </ul>
+   */
+  interface Map0Fun extends MapFun {
+
+    /**
+     * @param pageAlignedOffset It has to be a positive value that is page aligned.
+     */
+    MapSection map0(FileChannel fc, boolean readOnly, long pageAlignedOffset, long size)
```

Comment:
a beginner question: where is this method implemented?

Reply:
hahaha, good question!

There are 4 implementations of this interface: Map0Fun.ChronicleCore, Map0Fun.Java11, Map0Fun.Java17 and Map0Fun.Java20. All these implementations are lambda functions. After Java 8, interfaces with a single abstract method can be implemented by using an lambda with the same number of arguments and same return value. That is how, for example, you can implement a standard `java.util.function.Consumer` with a lambda. For example, a Consumer is defined as:

```
@FunctionalInterface
public interface Consumer<T> {

    /**
     * Performs this operation on the given argument.
     *
     * @param t the input argument
     */
    void accept(T t);

    /**
     * Returns a composed {@code Consumer} that performs, in sequence, this
     * operation followed by the {@code after} operation. If performing either
     * operation throws an exception, it is relayed to the caller of the
     * composed operation.  If performing this operation throws an exception,
     * the {@code after} operation will not be performed.
     *
     * @param after the operation to perform after this operation
     * @return a composed {@code Consumer} that performs in sequence this
     * operation followed by the {@code after} operation
     * @throws NullPointerException if {@code after} is null
     */
    default Consumer<T> andThen(Consumer<? super T> after) {
        Objects.requireNonNull(after);
        return (T t) -> { accept(t); after.accept(t); };
    }
}
```

Then you can say something like:

```
Consumer<Instant> aConsumer = (Instant i) -> System.out.println(i.toString())
```

and it is almost the same as writing:

```
Consumer<Instant> aConsumer = new Consumer<Instant> {
  void accept(Instant i) { 
      System.out.println(i.toString());
   }
};
```

It is not exactly the same because internally Java can do other optimizations. For example, in the case where we use `new Consumer<Instant>` Java needs to create a new instance each time the code executes that line, but in the lambda case the compiler can see that the lambda is not capturing and therefore create a single instance that will be shared between all executions.

BTW, you can see that Consumer interface declaration is decorated with `@FunctionalInterface`. This is an informative annotation that can be used to indicate that the interface is intended to be implemented by lambdas. It is not a requirement (and you can see I didn't use it in Map0Fun) but if added, the compiler can verify that the interface can actually be implemented as a lambda (in other words, it verifies that the interface has exactly one single not default method)